Sources/NIOTransportServices/FilterEmptyWritesHandler.swift

```diff
@@ -0,0 +1,76 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the SwiftNIO open source project
+//
+// Copyright (c) 2017-2018 Apple Inc. and the SwiftNIO project authors
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See CONTRIBUTORS.txt for the list of SwiftNIO project authors
+// swift-tools-version:4.0
+//
+// swift-tools-version:4.0
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+import NIO
+
+
+/// A `ChannelHandler` that checks for outbound writes of zero length, which are then dropped. This is
+/// due to a bug in `Network Framework`, where zero byte TCP writes lead to stalled connections.
+/// Write promises are confirmed in the correct order.
+public final class FilterEmptyWritesHandler: ChannelOutboundHandler {
+    public typealias OutboundIn = ByteBuffer
+    public typealias OutboundOut = ByteBuffer
+
+    private var prefixEmptyWritePromise: Optional<EventLoopPromise<Void>>
+    private var lastWritePromise: Optional<EventLoopPromise<Void>>
+    
+    public init() {
+        self.prefixEmptyWritePromise = nil
+        self.lastWritePromise = nil
+    }
+    
+    public func write(context: ChannelHandlerContext, data: NIOAny, promise: EventLoopPromise<Void>?) {
+        let buffer = self.unwrapOutboundIn(data)
+        if buffer.readableBytes > 0, let promise = promise {
+            self.lastWritePromise = promise
+        }
```

Comment:
if we don't get a promise here for a non-empty write, we need to create one. Otherwise, the following situation is handled wrongly:

```
write(1mb, nil)
write(0, promise)
flush()
```

the 0 write promise must only succeed when the previous write's done

Reply:
Changes pushed to create promise as suggested